### PR TITLE
[autograd] Enable dict inputs for torch.autograd.grad and torch.autograd.backward

### LINKED ISSUE #126650: Enable `dict` inputs for `torch.autograd.grad` and `torch.autograd.backward` (usability for `torch.func.functional_call`)
### 🚀 The feature, motivation and pitch

`functorch` is deprecated in favor of `torch.func` in PyTorch 2.0. The API [`functorch.make_functional`](https://pytorch.org/functorch/1.13/generated/functorch.make_functional.html#functorch.make_functional) is replaced by [`torch.func.functional_call`](https://pytorch.org/docs/stable/generated/torch.func.functional_call.html#torch.func.functional_call).

- #91811

> #### [`functorch.make_functional`](https://pytorch.org/docs/stable/func.migrating.html#functorch-make-functional)
>
> [torch.func.functional_call()](https://pytorch.org/docs/stable/generated/torch.func.functional_call.html#torch.func.functional_call) is the replacement for [functorch.make_functional](https://pytorch.org/functorch/1.13/generated/functorch.make_functional.html#functorch.make_functional) and [functorch.make_functional_with_buffers](https://pytorch.org/functorch/1.13/generated/functorch.make_functional_with_buffers.html#functorch.make_functional_with_buffers). However, it is not a drop-in replacement.

The `torch.func.functional_call()` API takes `dict[str, Tensor]` inputs for parameters and buffers, while `torch.autograd.{grad,backward}` only supports tensor and/or tuple of tensors (`Tensor | tuple[Tensor, ...]`) as input. This is very inconvenient.

This issue requests to support `dict[str, Tensor]` as inputs in `torch.autograd.grad` and `torch.autograd.backward`.

------

Code snippet for example case.

With `fmodel, params = functorch.make_functional(model)`, `params` is `tuple[nn.Parameter, ...]`:

```python
import functorch
import torch
import torch.nn as nn
import torch.nn.functional as F

fmodel: nn.Module
params: tuple[torch.Tensor, ...]
fmodel, params = functorch.make_functional(model)

def calculate_loss(fmodel, params, batch)
    inputs, labels = batch
    outputs = fmodel(params, inputs)
    return F.mse_loss(outputs, labels)

loss: torch.Tensor = calculate_loss(fmodel, params, batch)
grads: tuple[torch.Tensor, ...] = torch.autograd.grad(loss, params)
```

With `torch.func.functional_call(model, params)`, `params` is `dict[str, nn.Parameter]`:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

params: dict[str, torch.Tensor] = dict(model.named_parameters())

def calculate_loss(model, params, batch)
    inputs, labels = batch
    outputs = torch.func.functional_call(model, params, inputs)
    return F.mse_loss(outputs, labels)

loss: torch.Tensor = calculate_loss(model, params, batch)
grads: tuple[torch.Tensor, ...] = torch.autograd.grad(
    loss,
    tuple(params.values()),  # need manually convert `dict` to `tuple`
)
grads: dict[str, torch.Tensor] = dict(zip(params.keys(), grads))  # need manually convert `tuple` to `dict`
```

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #178500
* #178501
* __->__ #178140

Resolves #126650

Support `dict[str, Tensor]` (e.g. `dict(model.named_parameters())`) as the `inputs` argument to `torch.autograd.grad` and `torch.autograd.backward`.

When a dict is passed to `grad()`, the result is returned as a dict with matching keys. Other `Mapping` types (e.g. `OrderedDict`) are explicitly rejected with a `TypeError`; support for them is added in #178500.

This eliminates the manual dict-to-tuple and tuple-to-dict conversions previously required when using `torch.func.functional_call` with autograd.

### Core (`torch/autograd/__init__.py`)

- Detect `dict` inputs early, flatten to tuple of values for the C++ engine (unchanged), reconstruct dict from preserved keys on return.
- Add `@overload` signatures to narrow the return type of `grad()`: `dict` in → `dict` out; sequence in → `tuple` out.
- Explicitly reject non-`dict` `Mapping` types with `TypeError`.
- Unified empty-input validation with `RuntimeError` in both `grad()` and `backward()`, covering all input types (dicts, sequences, etc.). This changes the error type for `grad(y, [])` from `ValueError` (previously raised by the C++ engine) to `RuntimeError` (now caught at the Python level), consistent with `backward()`.

### Dynamo (`torch/_dynamo/variables/torch.py`, `torch/_dynamo/variables/tensor.py`)

- `handle_autograd_grad`: convert `ConstDictVariable` → `TupleVariable` before FX proxy creation; reconstruct `ConstDictVariable` after, reusing original `_HashableTracker` keys.
- `method_backward`: extract dict values from `ConstDictVariable` for `_collect_backward_inputs`.
- `_check_for_gradient_edge` / `_collect_tensors_with_sources`: recurse into dict values.

### Docstrings (`torch/_tensor.py`, `torch/autograd/__init__.py`)

- Update `inputs` parameter docs in `backward()`, `grad()`, and `Tensor.backward()` to mention dict support and clarify that `named_parameters()` needs to be wrapped with `dict()`.

### Tests

- 11 new eager tests in `test/test_autograd.py`: basic dict, `OrderedDict` rejection, `materialize_grads`, `allow_unused`, `create_graph`, `is_grads_batched`, non-string keys, empty dict, `backward()` with dict, `Tensor.backward()` with dict.
- 1 updated eager test: `test_grad_empty_inputs` now expects `RuntimeError` instead of `ValueError`.
- 3 Dynamo tests in `test/dynamo/test_fwd_loss_bwd.py`: `autograd.grad` with positional and keyword dict inputs, `loss.backward(inputs=dict(...))`.

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @Chillee @samdow @kshitij12345 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @azahed98

### CODE REVIEW

File: torch/autograd/__init__.py [hunk 1]
```diff
@@ -532,24 +577,31 @@ def vjp(gO):
...
         result = tuple(
             output
             if output is not None
-            else torch.zeros_like(input, requires_grad=create_graph)
-            for (output, input) in zip(result, inputs)
+            else torch.zeros_like(
+                input,  # pyrefly: ignore [bad-argument-type]
+                requires_grad=create_graph,
+            )
+            for (output, input) in zip(result, inputs_tuple)
```
Comment (Skylion007) [inline torch/autograd/__init__.py]: Strict kwarg please

File: torch/autograd/__init__.py [hunk 2]
```diff
@@ -532,24 +577,31 @@ def vjp(gO):
...
         )
+    if isinstance(inputs, Mapping):
+        if isinstance(inputs, OrderedDict):
```
Comment (Skylion007) [inline torch/autograd/__init__.py]: Strictness in these zips could be good too

File: torch/autograd/__init__.py [hunk 3]
```diff
@@ -467,20 +505,27 @@ def grad(
...
-    t_inputs = tuple(i for i in inputs if is_tensor_like(i))
+    t_inputs = tuple(i for i in inputs_tuple if is_tensor_like(i))
     overridable_args = t_outputs + t_inputs
     if has_torch_function(overridable_args):
         return handle_torch_function(
             grad,
             overridable_args,
             outputs,
-            inputs,
+            inputs,  # pyrefly: ignore [bad-argument-type]
```
Comment (Skylion007) [inline torch/autograd/__init__.py]: Add a cast/assert?

Comment (XuehaiPan) [reply torch/autograd/__init__.py]: This is due to pyrefly uses the first overload annotation.

File: torch/_dynamo/variables/torch.py [hunk 4]
```diff
@@ -277,6 +279,14 @@ def _check_for_gradient_edge(var: VariableTracker, arg_name: str) -> None:
     elif isinstance(var, BaseListVariable):
         for i, item in enumerate(var.items):
             _check_for_gradient_edge(item, f"{arg_name}[{i}]")
+    elif isinstance(var, ConstDictVariable):
+        for hash_key, item in var.items.items():
+            key_repr = (
+                hash_key.vt.as_python_constant()
+                if hash_key.vt.is_python_constant()
+                else "?"
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/torch.py]: "?" seems wrong here - especially because if we have multiple key failures we will get this same repr... Should we error loudly instead?

File: torch/_tensor.py [hunk 5]
```diff
@@ -613,10 +613,12 @@ def backward(
             create_graph (bool, optional): If ``True``, graph of the derivative will
                 be constructed, allowing to compute higher order derivative
                 products. Defaults to ``False``.
-            inputs (Sequence[Tensor], optional): Inputs w.r.t. which the gradient will be
-                accumulated into ``.grad``. All other tensors will be ignored. If not
-                provided, the gradient is accumulated into all the leaf Tensors that were
-                used to compute the :attr:`tensors`. Defaults to ``None``.
+            inputs (Sequence[Tensor] or dict[str, Tensor], optional): Inputs w.r.t. which
```
Comment (Lucaskabela) [inline torch/_tensor.py:616]: nit: use "Mapping" here too

**Review by Lucaskabela — DISMISSED** [review]
Some thoughts on our fallback strategy here

**Review by zou3519 — COMMENTED** [review]
Please let @soulitzer and @albanD take a look at this. The last time we discussed, this was explicitly out of scope for these two apis

File: torch/_tensor.py [hunk 6]
```diff
@@ -613,10 +613,12 @@ def backward(
...
                 products. Defaults to ``False``.
-            inputs (Sequence[Tensor], optional): Inputs w.r.t. which the gradient will be
-                accumulated into ``.grad``. All other tensors will be ignored. If not
-                provided, the gradient is accumulated into all the leaf Tensors that were
-                used to compute the :attr:`tensors`. Defaults to ``None``.
+            inputs (Sequence[Tensor] or Mapping[str, Tensor], optional): Inputs w.r.t. which
+                the gradient will be accumulated into ``.grad``. All other tensors will be
+                ignored. If not provided, the gradient is accumulated into all the leaf
+                Tensors that were used to compute the :attr:`tensors`. A dict of tensors
+                (e.g. from :meth:`~torch.nn.Module.named_parameters`) is also accepted.
```
Comment (soulitzer) [inline torch/_tensor.py]: Could this be sentence be confusing since `torch.nn.Module.named_parameters` doesn't actually return a dict directly. Should we mention that the user call dict(...) on the result.

File: torch/autograd/__init__.py [hunk 7]
```diff
@@ -532,24 +587,30 @@ def vjp(gO):
...
+            for (r, inp) in zip(
+                result,
+                cast(tuple[torch.Tensor, ...], inputs_tuple),
+                strict=True,
+            )
         )
+    if isinstance(inputs, Mapping):
+        if isinstance(inputs, OrderedDict):
+            return OrderedDict(zip(inputs.keys(), result, strict=True))
+        return dict(zip(inputs.keys(), result, strict=True))
```
Comment (soulitzer) [inline torch/autograd/__init__.py:620]: Could it be confusing if user passes in their Mapping and only a dict is returned? Let's start with supporting just "dict" and explicitly error for any other mapping type.

File: torch/autograd/__init__.py [hunk 8]
```diff
@@ -532,24 +587,30 @@ def vjp(gO):
...
+            r if r is not None else torch.zeros_like(inp, requires_grad=create_graph)
+            for (r, inp) in zip(
+                result,
+                cast(tuple[torch.Tensor, ...], inputs_tuple),
+                strict=True,
+            )
         )
+    if isinstance(inputs, Mapping):
+        if isinstance(inputs, OrderedDict):
+            return OrderedDict(zip(inputs.keys(), result, strict=True))
```
Comment (soulitzer) [inline torch/autograd/__init__.py]: Not sure we need to support the OrderedDict case. state_dict indeed returns a OrderedDict by default it detaches and returns buffers as well, so not actually sure its very useful for that.

File: test/test_autograd.py [hunk 9]
```diff
@@ -1725,7 +1726,7 @@ def test_grad_badcalls(self):
 
     def test_grad_empty_inputs(self):
         x = torch.tensor([1.0], requires_grad=True)
-        with self.assertRaisesRegex(ValueError, "grad requires non-empty inputs."):
+        with self.assertRaisesRegex(RuntimeError, "cannot be empty"):
```
Comment (soulitzer) [inline test/test_autograd.py:1730]: Technically bc-breaking but fine since its probably rare and this makes it consistent between .backward and .grad
Although now that we're expecting catching the error in python instead of the engine, we could make the error in the engine an internal assert to make it clearer.

File: torch/_tensor.py (hunk 6 above)
Comment (XuehaiPan) [reply torch/_tensor.py]: Updated. Changed to `dict(model.named_parameters())` to clarify that `named_parameters()` returns an iterator, not a dict directly.

File: torch/autograd/__init__.py (hunk 7 above)
Comment (XuehaiPan) [reply torch/autograd/__init__.py:620]: The previous state of the `backward`/`grad` functions allows users to pass `Sequence[Tensor]` which always casts them to `tuple`. A user can pass a `list` and get a `tuple` back. I think it is OK for us to have a narrowed return type here as well. I changed to reject `Mapping` types other than `dict` with a `TypeError` and then enable them in #178500.

File: torch/autograd/__init__.py (hunk 8 above)
Comment (XuehaiPan) [reply torch/autograd/__init__.py]: Moved to #178500. This PR now only allows `dict`.

File: test/test_autograd.py (hunk 9 above)
Comment (XuehaiPan) [reply test/test_autograd.py:1730]: Opened #178501.

**Review by soulitzer — APPROVED** [review]
Thanks for the update, and sorry for the delay! The eager side looks good to me. But  someone on the dynamo side should review the compile support parts.

**Comment by XuehaiPan** [comment]
ping @Lucaskabela for code review, thanks!

**Review by anijain2305 — COMMENTED** [review]
Dynamo changes LGTM!

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting but it seems to cause some trunk failures' -c nosignal

test/inductor/test_compiled_autograd.py::TestAutogradWithCompiledAutograd::test_grad_dict_inputs_batched_grads [GH job link](https://github.com/pytorch/pytorch/actions/runs/25239451614/job/74016695652) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/7b5f32b1c4911f959ed9f61cd0aefb7ed57e0317)
